drivers/display/display_nrf_led_matrix.c

```diff
@@ -0,0 +1,448 @@
+
+#if USE_PWM
+	uint32_t out_psels[NRF_PWM_CHANNEL_COUNT] = {
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+		NRF_PWM_PIN_NOT_CONNECTED,
+	};
+	nrf_pwm_sequence_t sequence = {
+		.values.p_raw = &dev_data->seq,
+		.length = 1,
+	};
+
+	nrf_pwm_pins_set(dev_config->pwm, out_psels);
+	BUILD_ASSERT(PIXEL_PERIOD <= PWM_COUNTERTOP_COUNTERTOP_Msk);
+	nrf_pwm_configure(dev_config->pwm,
+		PWM_CLK_CONFIG, NRF_PWM_MODE_UP, PIXEL_PERIOD);
+	nrf_pwm_decoder_set(dev_config->pwm,
+		NRF_PWM_LOAD_COMMON, NRF_PWM_STEP_TRIGGERED);
+	nrf_pwm_sequence_set(dev_config->pwm, 0, &sequence);
+	nrf_pwm_loop_set(dev_config->pwm, 0);
+	nrf_pwm_shorts_set(dev_config->pwm, NRF_PWM_SHORT_SEQEND0_STOP_MASK);
+#else
+	nrfx_err_t err;
+	nrf_ppi_channel_t ppi_ch;
+
+	err = nrfx_ppi_channel_alloc(&ppi_ch);
+	if (err != NRFX_SUCCESS) {
+		return -ENOMEM;
+	}
+
+	err = nrfx_gpiote_channel_alloc(&dev_data->gpiote_ch);
+	if (err != NRFX_SUCCESS) {
+		nrfx_ppi_channel_free(ppi_ch);
+		return -ENOMEM;
+	}
+
+	nrf_ppi_channel_endpoint_setup(NRF_PPI, ppi_ch,
+		nrf_timer_event_address_get(dev_config->timer,
+			nrf_timer_compare_event_get(1)),
+		nrf_gpiote_event_address_get(NRF_GPIOTE,
+			nrf_gpiote_out_task_get(dev_data->gpiote_ch)));
+	nrf_ppi_channel_enable(NRF_PPI, ppi_ch);
+#endif /* USE_PWM */
+
+	for (uint8_t i = 0; i < ROW_COUNT; ++i) {
+		uint8_t row_pin_info = dev_config->rows[i];
+
+		set_pin(row_pin_info, false);
+		nrf_gpio_cfg(row_pin_info & PSEL_MASK,
+			     NRF_GPIO_PIN_DIR_OUTPUT,
+			     NRF_GPIO_PIN_INPUT_DISCONNECT,
+			     NRF_GPIO_PIN_NOPULL,
+			     NRF_GPIO_PIN_S0S1,
+			     NRF_GPIO_PIN_NOSENSE);
+	}
+
+	for (uint8_t i = 0; i < COL_COUNT; ++i) {
+		uint8_t col_pin_info = dev_config->cols[i];
+
+		set_pin(col_pin_info, false);
+		nrf_gpio_cfg(col_pin_info & PSEL_MASK,
+			     NRF_GPIO_PIN_DIR_OUTPUT,
+			     NRF_GPIO_PIN_INPUT_DISCONNECT,
+			     NRF_GPIO_PIN_NOPULL,
+			     NRF_GPIO_PIN_S0S1,
+			     NRF_GPIO_PIN_NOSENSE);
+	}
+
+	nrf_timer_bit_width_set(dev_config->timer, NRF_TIMER_BIT_WIDTH_16);
+	nrf_timer_frequency_set(dev_config->timer, TIMER_CLK_CONFIG);
+	nrf_timer_cc_set(dev_config->timer, 0, PIXEL_PERIOD);
+	nrf_timer_shorts_set(dev_config->timer,
+			     NRF_TIMER_SHORT_COMPARE0_STOP_MASK |
+			     NRF_TIMER_SHORT_COMPARE0_CLEAR_MASK);
+	nrf_timer_event_clear(dev_config->timer, NRF_TIMER_EVENT_COMPARE0);
+	nrf_timer_int_enable(dev_config->timer, NRF_TIMER_INT_COMPARE0_MASK);
+
+	IRQ_CONNECT(DT_IRQN(TIMER_NODE), DT_IRQ(TIMER_NODE, priority),
+		    timer_irq_handler, DEVICE_DT_GET(MATRIX_NODE), 0);
+	irq_enable(DT_IRQN(TIMER_NODE));
+
+	return 0;
+}
+
+static struct display_drv_data instance_data = {
+	.brightness = 0xFF,
+	.blanking   = true,
+};
+
+#define GET_PIN_INFO(node_id, pha, idx) \
+	(DT_GPIO_PIN_BY_IDX(node_id, pha, idx) | \
+	 (DT_PROP_BY_PHANDLE_IDX(node_id, pha, idx, port) << 5) | \
+	 ((DT_GPIO_FLAGS_BY_IDX(node_id, pha, idx) & GPIO_ACTIVE_LOW) ? \
+		ACTIVE_LOW_MASK : 0)),
```

Comment:
ditto for GPIO stuff, gpio_dt_spec and GPIO API may be used

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/40968#discussion_r771270367.